File: src/operations.c
#include "calculator.h"

double add(double a, double b) {
    return a + b;
}

double subtract(double a, double b) {
    return a - b;
}

double multiply(double a, double b) {
    return a * b;
}

double divide(double a, double b) {
    if (b == 0) {
        fprintf(stderr, "Error: Division by zero\n");
        return 0;
    }
    return a / b;
}

double power(double a, double b) {
    return pow(a, b);
}

double modulo(double a, double b) {
    if (b == 0) {
        fprintf(stderr, "Error: Modulo by zero\n");
        return 0;
    }
    return fmod(a, b);
}

double floor_divide(double a, double b) {
    if (b == 0) {
        fprintf(stderr, "Error: Floor division by zero\n");
        return 0;
    }
    return floor(a / b);
}

double maximum(double a, double b) {
    return (a > b) ? a : b;
}

double minimum(double a, double b) {
    return (a < b) ? a : b;
}

double square(double a) {
    return a * a;
}

double cube(double a) {
    return a * a * a;
}

double negate(double a) {
    return -a;
}

double absolute(double a) {
    return fabs(a);
}

double square_root(double a) {
    if (a < 0) {
        fprintf(stderr, "Error: Cannot take square root of negative number\n");
        return 0;
    }
    return sqrt(a);
}

double logarithm(double a) {
    if (a <= 0) {
        fprintf(stderr, "Error: Cannot take logarithm of non-positive number\n");
        return 0;
    }
    return log(a);
}

double exponential(double a) {
    return exp(a);
}

double sine(double a) {
    return sin(a);
}

double cosine(double a) {
    return cos(a);
}

double tangent(double a) {
    double c = cos(a);
    if (c == 0) {
        fprintf(stderr, "Error: Tangent undefined at this angle\n");
        return 0;
    }
    return tan(a);
}

double hyperbolic_tangent(double a) {
    return tanh(a);
}

double relu(double a) {
    return (a > 0) ? a : 0;
}

double sigmoid(double a) {
    return 1.0 / (1.0 + exp(-a));
}

int is_integer(double value) {
    return value == floor(value);
}

char *format_result(double value, char *buffer, size_t size) {
    if (is_integer(value) && fabs(value) < 1e15) {
        snprintf(buffer, size, "%.0f", value);
    } else {
        snprintf(buffer, size, "%.10g", value);
    }
    return buffer;
}

File: calculator/operations.py
"""
Core arithmetic operations for the calculator package.
"""


def add(a: float, b: float) -> float:
    """Return the sum of a and b.

    Args:
        a: The first number.
        b: The second number.

    Returns:
        The result of a + b.
    """
    return a + b


def subtract(a: float, b: float) -> float:
    """Return the difference of a and b.

    Args:
        a: The first number.
        b: The second number.

    Returns:
        The result of a - b.
    """
    return a - b


def multiply(a: float, b: float) -> float:
    """Return the product of a and b.

    Args:
        a: The first number.
        b: The second number.

    Returns:
        The result of a * b.
    """
    return a * b


def divide(a: float, b: float) -> float:
    """Return the quotient of a and b.

    Args:
        a: The dividend.
        b: The divisor.

    Returns:
        The result of a / b.

    Raises:
        ZeroDivisionError: If b is zero.
    """
    if b == 0:
        raise ZeroDivisionError("Cannot divide by zero.")
    return a / b


def power(a: float, b: float) -> float:
    """Return a raised to the power b.

    Args:
        a: The base.
        b: The exponent.

    Returns:
        The result of a ** b.
    """
    return a ** b


def modulo(a: float, b: float) -> float:
    """Return the remainder of a divided by b.

    Args:
        a: The dividend.
        b: The divisor.

    Returns:
        The result of a % b.

    Raises:
        ZeroDivisionError: If b is zero.
    """
    if b == 0:
        raise ZeroDivisionError("Cannot modulo by zero.")
    return a % b


def floor_divide(a: float, b: float) -> float:
    """Return the floor division of a by b.

    Args:
        a: The dividend.
        b: The divisor.

    Returns:
        The result of a // b.

    Raises:
        ZeroDivisionError: If b is zero.
    """
    if b == 0:
        raise ZeroDivisionError("Cannot floor divide by zero.")
    return a // b


def maximum(a: float, b: float) -> float:
    """Return the larger of a and b.

    Args:
        a: The first number.
        b: The second number.

    Returns:
        The maximum of a and b.
    """
    return max(a, b)


def minimum(a: float, b: float) -> float:
    """Return the smaller of a and b.

    Args:
        a: The first number.
        b: The second number.

    Returns:
        The minimum of a and b.
    """
    return min(a, b)


def square(a: float) -> float:
    """Return the square of a.

    Args:
        a: The number to square.

    Returns:
        The result of a².
    """
    return a ** 2


def cube(a: float) -> float:
    """Return the cube of a.

    Args:
        a: The number to cube.

    Returns:
        The result of a³.
    """
    return a ** 3


def negate(a: float) -> float:
    """Return the negation of a.

    Args:
        a: The number to negate.

    Returns:
        The result of -a.
    """
    return -a


def absolute(a: float) -> float:
    """Return the absolute value of a.

    Args:
        a: The number.

    Returns:
        The absolute value of a.
    """
    return abs(a)


def square_root(a: float) -> float:
    """Return the square root of a.

    Args:
        a: The number.

    Returns:
        The square root of a.

    Raises:
        ValueError: If a is negative.
    """
    if a < 0:
        raise ValueError("Cannot take square root of a negative number.")
    return a ** 0.5


def logarithm(a: float) -> float:
    """Return the natural logarithm of a.

    Args:
        a: The number.

    Returns:
        The natural log (ln) of a.

    Raises:
        ValueError: If a is negative or zero.
    """
    if a <= 0:
        raise ValueError("Cannot take logarithm of non-positive number.")
    import math
    return math.log(a)


def exponential(a: float) -> float:
    """Return e raised to the power of a.

    Args:
        a: The exponent.

    Returns:
        The result of e^a.
    """
    import math
    return math.exp(a)


def sine(a: float) -> float:
    """Return the sine of a (in radians).

    Args:
        a: The angle in radians.

    Returns:
        The sine of a.
    """
    import math
    return math.sin(a)


def cosine(a: float) -> float:
    """Return the cosine of a (in radians).

    Args:
        a: The angle in radians.

    Returns:
        The cosine of a.
    """
    import math
    return math.cos(a)


def tangent(a: float) -> float:
    """Return the tangent of a (in radians).

    Args:
        a: The angle in radians.

    Returns:
        The tangent of a.

    Raises:
        ValueError: If cosine is zero.
    """
    import math
    return math.tan(a)


def hyperbolic_tangent(a: float) -> float:
    """Return the hyperbolic tangent of a.

    Args:
        a: The number.

    Returns:
        The hyperbolic tangent of a.
    """
    import math
    return math.tanh(a)


def relu(a: float) -> float:
    """Return the Rectified Linear Unit of a.

    Args:
        a: The number.

    Returns:
        max(0, a)
    """
    return max(0, a)


def sigmoid(a: float) -> float:
    """Return the sigmoid function of a.

    Args:
        a: The number.

    Returns:
        1 / (1 + e^(-a))
    """
    import math
    return 1 / (1 + math.exp(-a))
